turbopack/crates/turbo-tasks-backend/src/data.rs

```diff
@@ -83,6 +85,140 @@ transient_traits!(RootState);
 
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct DirtyState {
+    pub clean_in_session: Option<SessionId>,
+}
+
+impl DirtyState {
+    pub fn get(&self, session: SessionId) -> bool {
+        self.clean_in_session != Some(session)
+    }
+}
+
+fn add_with_diff(v: &mut i32, u: i32) -> i32 {
+    let old = *v;
+    *v += u;
+    if old <= 0 && *v > 0 {
+        1
+    } else if old > 0 && *v <= 0 {
+        -1
+    } else {
+        0
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct DirtyContainerCount {
+    pub count: i32,
+    pub count_in_session: Option<(SessionId, i32)>,
+}
+
+impl DirtyContainerCount {
+    pub fn get(&self, session: SessionId) -> i32 {
+        if let Some((s, count)) = self.count_in_session {
+            if s == session {
+                return count;
+            }
+        }
+        self.count
+    }
+
+    pub fn update(&mut self, count: i32) -> DirtyContainerCount {
+        self.update_count(&DirtyContainerCount {
+            count,
+            count_in_session: None,
+        })
+    }
+
+    pub fn update_session_dependent(
+        &mut self,
+        ignore_session: SessionId,
+        count: i32,
+    ) -> DirtyContainerCount {
+        self.update_count(&DirtyContainerCount {
+            count,
+            count_in_session: Some((ignore_session, 0)),
+        })
+    }
+
+    pub fn update_count(&mut self, count: &DirtyContainerCount) -> DirtyContainerCount {
+        let mut diff = DirtyContainerCount::default();
+        match (
+            self.count_in_session.as_mut(),
+            count.count_in_session.as_ref(),
+        ) {
+            (None, None) => {}
+            (Some((s, c)), None) => {
+                let d = add_with_diff(c, count.count);
+                diff.count_in_session = Some((*s, d));
+            }
+            (None, Some((s, c))) => {
+                let mut new = self.count;
+                let d = add_with_diff(&mut new, *c);
+                self.count_in_session = Some((*s, new));
+                diff.count_in_session = Some((*s, d));
+            }
+            (Some((s1, c1)), Some((s2, c2))) => match (*s1).cmp(s2) {
+                Ordering::Less => {
+                    let mut new = self.count;
+                    let d = add_with_diff(&mut new, *c2);
+                    self.count_in_session = Some((*s2, new));
+                    diff.count_in_session = Some((*s2, d));
+                }
+                Ordering::Equal => {
+                    let d = add_with_diff(c1, *c2);
+                    diff.count_in_session = Some((*s1, d));
+                }
+                Ordering::Greater => {
+                    let d = add_with_diff(c1, count.count);
+                    diff.count_in_session = Some((*s1, d));
+                }
+            },
+        }
+        let d = add_with_diff(&mut self.count, count.count);
+        diff.count = d;
+        diff
+    }
```

Comment:
I'm not sure about the best way to fix this (a distinct type that uses an enum instead of an `i32`?), but re-using `DirtyContainerCount` for diffs is a bit confusing. At the very least, I feel like some of these methods could use docs.